docs/architecture/adr-081-protobuf-mgmt.md

```diff
@@ -0,0 +1,202 @@
+# ADR 081: Protocol Buffers Management
+
+## Changelog
+
+- 2022-02-28: First draft
+
+## Status
+
+Accepted
+
+## Context
+
+At present, we manage the [Protocol Buffers] files ("protos") that define our
+wire-level data formats within the Tendermint repository itself (see the
+[`proto`](../../proto/) directory). Recently, we have been making use of [Buf],
+both locally and in CI, in order to generate Go stubs, and lint and check
+`.proto` files for breaking changes.
+
+The version of Buf used at the time of this decision was `v1beta1`, and it was
+discussed in [\#7975] and in weekly calls as to whether we should upgrade to
+`v1` and harmonize our approach with that used by the Cosmos SDK. The team
+managing the Cosmos SDK was primarily interested in having our protos versioned
+and easily accessible from the [Buf] registry.
+
+The three main customers for the `.proto` files and their needs, as currently
+understood, are as follows.
+
+1. Tendermint needs Go code generated from `.proto` files.
+2. Consumers of Tendermint's `.proto` files want to be able to access these
+   files in a reliable and efficient way.
+3. The Tendermint Core team wants to provide stable interfaces that are as easy
+   as possible to maintain, on which consumers can depend, and to be able to
+   notify those consumers promptly when those interfaces change. To this end, we
+   want to:
+   1. Prevent any breaking changes from being introduced in minor/patch releases
+      of Tendermint. Only major version updates should be able to contain
+      breaking interface changes.
+   2. Ensure that `.proto` files don't contain invalid syntax.
```

Comment:
I think more important than syntax alone is to ensure that the Go generated code that is used by the Tendermint core implementation is up-to-date. That subsumes correct syntax, but I feel like there shouldn't be any changes to the production proto files that aren't immediately reflected in the implementation.

That does mean if we are making incompatible changes we might need to stage them (perhaps, as @williambanfield has suggested, in a branch; or in a "devel" directory, or whatever). But that should be comparatively rare even on master.

Reply:
> > That does mean if we are making incompatible changes we might need to stage them
> 
> Would you mind elaborating on why that may be? I _think_ we can generally use a 'strangler' pattern for our proto changes, even for breaking changes

Yes, some breaking changes could be staged in "breaking" and "non-breaking" phases. I'm just saying that if we decide to make bigger changes, e.g., Tendermint 2.0 and ABCI-TheNextGeneration, we might want to make a larger wholesale change to the protos.

In such a (hypothetical) case, I think we'd want to make those changes in separate files, since presumably fixing up the build would not be straightforward. But I don't see this as being the common case.

Reply:
> In such a (hypothetical) case, I think we'd want to make those changes in separate files, since presumably fixing up the build would not be straightforward. But I don't see this as being the common case.

I see, that's possible. I'd argue that the correct relationship for staged protos to the repo would be in some combination of ADR + prototype / demo branch.

Reply:
Yes, I agree. I certainly don't think we need to use a next-version staging file for normal development.